### PR TITLE
fix: selection highlight

### PR DESCRIPTION
**Pull request by lumixraku**
close https://github.com/dream-num/univer-pro/issues/3020
close https://github.com/dream-num/univer-pro/issues/2791
close https://github.com/dream-num/univer-pro/issues/2204
close https://github.com/dream-num/univer-pro/issues/2896
close https://github.com/dream-num/univer/issues/3868

move 
move addLinkToDocumentModel from sheets-ui to core

rename
WorkbookSelections ---> WorkbookSelectionDataModel and moved to new file
getCellByIndexWithNoHeader  ---> getCellWithCoordByIndex  (old one still exists)
ISelectionCellWithMergeInfo --->  ICellWithCoord   (old one is kept)
tools@getCellByIndexWithMergeInfo ---> getCellWithCoordByIndexCore  (original name is kept)
getFormulaRefSelectionStyle ---> genFormulaRefSelectionStyle

base-selection-render.service.ts@getSelectionCellByPosition ---> base-selection-render.service.ts@getCellWithCoordByOffset  

sk@calculateCellIndexByPosition ---> sk@getCellWithCoordByOffset
sk@getCellPositionByOffset ---> sk@getCellIndexByOffset
sk@getColumnPositionByOffsetX --->  sk@getColumnIndexByOffsetX
sk@getCellByIndexWithNoHeader(row, col) --->  getCellWithCoordByIndex(row, col, false);
sk@getCellByIndex ---> getCellWithCoordByIndex    (Old one still exists)

scene@getRelativeToViewportCoord --->   getCoordRelativeToViewport  (origin name still exists)
scene@getVpScrollXYInfoByPosToVp --->  getScrollXYInfoByViewport (origin name still exists)

makeCellToSelection  ---> convertCellToRange  (origin function still exists)
getTopLeftSelection ---> getTopLeftSelectionOfCurrSheet   (origin function still exists)

remove
engine-render/src/components/sheets/extensions/sheet-extension.ts@getCellByIndex, no one use this method.

remove useless props for selection style,  this should be controlled by props in selection control.
hasAutoFill
hasRowHeader
hasColumnHeader

remove defaultStyle for selectionControl, only keep currentStyle. 

rewrite function because rm deprecated function
packages/sheets-ui/src/services/selection/base-selection-render.service.ts@getTopLeftSelection

## Pull Request Checklist

- [ ] Related tickets or issues have been linked in the PR description (or missing issue).
- [ ] [Naming convention](https://github.com/dream-num/univer/blob/dev/docs/NAMING_CONVENTION.md) is followed (**do please** check it especially when you created new plugins, commands and resources).
- [ ] Unit tests have been added for the changes (if applicable).
- [ ] Breaking changes have been documented (or no breaking changes introduced in this PR).

### CODE REVIEW

**Comment by oumomomo** [comment]
https://github.com/dream-num/univer-pro/issues/3020
反复拖拽范围选区，未选中的区域展示了选中颜色
image
[2024_11_6 11_35_01 video.webm](https://github.com/user-attachments/assets/16dbd545-3b5b-4119-b88e-c23b7e747d37)

**Comment by oumomomo** [comment]
https://github.com/dream-num/univer-pro/issues/2896
反复拖拽选区，选区在最后定位时会跳动到其他区域
[2024_11_6 19_28_25 video.webm](https://github.com/user-attachments/assets/4a064b71-7865-45ae-984e-8e745f40551d)

File: packages/core/src/sheets/typedef.ts [hunk 1]
```diff
@@ -594,12 +594,75 @@ export interface ISingleCell {
...
+ * startColumn: number;
+ * endRow: number;
+ * endColumn: number;
+ *
+ * actualRow: number;
+ * actualColumn: number;
+ * isMerged: boolean;
+ * isMergedMainCell: boolean;
+ */
+// Bad name!! selection is a concept in sheet-ui !!
```
Comment (wzhudev) [inline packages/core/src/sheets/typedef.ts:661]: Bad name? Yes. Concept in sheet-ui? Not quite if you take Node.js SDK into consideration.

**Review by wzhudev — APPROVED** [review]

File: e2e/disposing/disposing.spec.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ test('no error on constructing and disposing sheet unit', async ({ page }) => {
         errored = true;
     });
 
-    await page.goto('http://localhost:3000/sheets/', { waitUntil: 'networkidle' });
```
Comment (Gggpound) [inline e2e/disposing/disposing.spec.ts:50]: 为什么要改这里

File: examples-node/src/locales.ts [hunk 3]
```diff
@@ -17,45 +17,57 @@
...
+import sheetsformulazhTW from '@univerjs/sheets-formula/locale/zh-TW';
+import sheetsformulaviVN from '@univerjs/sheets-formula/locale/vi-VN';
+import sheetsformulafaIR from '@univerjs/sheets-formula/locale/fa-IR';
 import sheetssortenUS from '@univerjs/sheets-sort/locale/en-US';
-import sheetssortfaIR from '@univerjs/sheets-sort/locale/fa-IR';
 import sheetssortruRU from '@univerjs/sheets-sort/locale/ru-RU';
-import sheetssortviVN from '@univerjs/sheets-sort/locale/vi-VN';
 import sheetssortzhCN from '@univerjs/sheets-sort/locale/zh-CN';
 import sheetssortzhTW from '@univerjs/sheets-sort/locale/zh-TW';
+import sheetssortviVN from '@univerjs/sheets-sort/locale/vi-VN';
```
Comment (Gggpound) [inline examples-node/src/locales.ts:35]: 这个pr不应该涉及这个改动吧？

File: e2e/disposing/disposing.spec.ts (hunk 2 above)
Comment (lumixraku) [reply e2e/disposing/disposing.spec.ts:50]: ok  revert

File: examples-node/src/locales.ts (hunk 3 above)
Comment (lumixraku) [reply examples-node/src/locales.ts:35]: lint autofix because of  import sort rule

File: packages/core/src/sheets/typedef.ts (hunk 1 above)
Comment (lumixraku) [reply packages/core/src/sheets/typedef.ts]: Still a bad name even node needs it. But ICellInfo isn't much better either. 
This comments is deleted, Both interfaces kept.

File: examples-node/src/locales.ts (hunk 3 above)
Comment (lumixraku) [reply examples-node/src/locales.ts]: fixed

**Pull request merged**